tonilopezmr/day 3/SpiralMemory.kt

```diff
@@ -0,0 +1,109 @@
+package day3
+
+import org.junit.Assert.*
+import org.junit.Test
+
+fun Int.pow(exponent: Int): Double = Math.pow(this.toDouble(), exponent.toDouble())
+fun Int.abs(): Int = Math.abs(this)
+fun Double.floor(): Int = Math.floor(this).toInt()
+
+typealias Dimens = Pair<Int, Int>
+
+class SpiralMemory {
+
+  private fun getCenter(spiralDimension: Int): Dimens {
+    val dimension = (spiralDimension.toDouble() / 2.0).floor()
+    return Dimens(dimension, dimension)
+  }
+
+  private fun manhatan(a: Dimens, b: Dimens): Int {
+    return (a.first - b.first).abs() + (a.second - b.second).abs()
+  }
+
+  private fun spiralDimens(input: Int): Int {
+    var base = 2
+    while (base.pow(2) < input) base++
+    return base
+  }
+
+  fun spiralMemory(matrix: Int, center: Dimens, input: Int): Int {
+    var displacementX = 0
+    var displacementY = -1
+    var x = 0
+    var y = 0
+    var number = 1
+
+    while (number <= matrix.pow(2).toInt()) {
+      val realX = center.first + x
+      val realY = center.second + y
+
+      if (input == number) {
+        return manhatan(center, Dimens(realX, realY))
+      }
+
+      if (x == y || (x < 0 && x == -y) || (x > 0 && x == 1 - y)) {
```

Comment:
How is it possible for `x` to be Negative? Do you consider (0,0) as the center of your matrix? I don't get it 😖

Reply:
Yes! 0,0 is the center

Reply:
Yes I iterate with a **displacement** like this, (0,0) is 1, (-1,0) is 2, (0,-1) is 3.. and go on..

But the problem here is that you need to know the center, ej: (5, 5), and always sum the **desplacement**, but with *Dictionaries* like @JoseLlorensRipolles solution, you don't need to know the center, or somethink like that because you don't care about BoundExceptions.. You only need a position for 1, 2, 3..